uiowahjmjohnsonselt2023/selt2023-final-project-team-001
# Sign in

# Pull Request: Sign-In Feature

## Description

This pull request adds a sign-in feature to the project. Users can now log in to access their accounts.

## Changes Made

Login Changes:
- Added new sign-in functionality.
- Add view for /login with a barebones login form
- Added routes for /login and /logout
- Added controller for logging in and out, session_controller
- Added testing for session_controller
- Added testing for the session requests HTTP responses
- Flash messages for correct and incorrect logins as well as logging out

General Changes:
- Added coverage folder to .gitignore
- Added a user factory for ease of testing
- Pulled an early version of the create-user branch for testing purposes
- Updated seeding of users to use password instead of password_digest
- Add flash notice and flash alert to application.html.erb so all pages will show flash messages, also some styling added for those flash messages
- Added a dummy home page and routing to test my flash messages after redirect

## Usage and Testing

### To manually test the sign-in feature:
Happy Path:
1. Visit '/login' in your local development environment.
2. Enter valid user credentials after seeding the database.
3. Click the "Log In" button.
4. Verify that you are redirected to '/'.
Sad Path:
1.  Visit '/login' in your local development environment.
2. Enter invalid user credentials after seeding the database.
3. Click the "Log In" button.
4. Verify that you are not redirected to '/' and instead stay on the page.

#### To run the rspec tests put this in your terminal
``` bundle exec rspec spec/controllers/sessions_controller_spec.rb```

## Impact on Existing Code
- No breaking changes are introduced.
- User seeding needed to change

## Screenshots or Demos
Sad Path login:
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/e7b0e903-5359-477a-85cf-2fb5caf8474b)
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/59faa96b-2dd6-4d6a-ba57-be2484088928)

Happy Path login:
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/88c11259-10e5-4f95-9462-0993583107c6)
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/3b96a931-bd2d-4937-b901-dc68511d1545)

Sad Path logout:
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/e906c913-4425-46db-a779-c8e59a7e65bc)
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/85c6afb4-591f-4ebc-bc5d-1e746fa9e6c4)

Happy Path logout:
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/8dd618d2-a36b-430e-98a8-ee9b1ce5ae09)
![image](https://github.com/uiowahjmjohnsonselt2023/selt2023-final-project-team-001/assets/112444260/88905400-ae58-45f9-bce0-878152fcb6c8)





## Additional Notes

- This feature is part of the ongoing effort to improve user experience and security.
- No major performance issues have been observed during testing.
- If you seeded the database with the previous seed that went straight to password_digest and try to login as admin@admin.com; admin you will encounter a bug because it was improperly seeded. To fix, on my branch run
```
bin/rails db:drop
bin/rails db:create
bin/rails db:migrate
bin/rails db:seed
```

## Reviewer Requests

I kindly request a review of these changes. Please provide feedback and suggestions. As this was a 2 point user story just 1 review should suffice.

## Status Checks and Tests

- All automated tests have passed.
- Code coverage is ```Coverage report generated for RSpec to /mnt/c/Users/blake/selt2023-final-project-team-001/coverage. 167 / 168 LOC (99.4%) covered.```
- No known issues or limitations at this time.

## Discussion

Feel free to comment, ask questions, or provide feedback.
